msyslog-v1.08a+smac/src/man/om_regex.8
.\"	$CoreSDI: om_regex.8,v 1.1.2.7.2.1.4.10 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Aug 20, 2000
.Dt OM_REGEXP 8
.Os Core-SDI
.Sh NAME
.Nm msyslog regex output module
.Nd
.Xr syslogd 8
output module used to filter and redirect logs
.Sh SYNOPSIS
.Nm regex
.Op Fl v
.Op Fl mhdt Ar exp
.Sh DESCRIPTION
.Nm regex output module 
receives a message and checks whether the regular expression
.Ar exp
matches given field of a log. You must select ONE of the following parameters:
.Bl -tag -width Ds
.It Fl v
reverse result
.It Fl m Ar exp
match exp in message
.It Fl h Ar exp
match exp in host
.It Fl d Ar exp
match exp in date
.It Fl t Ar exp
match exp in time
.El
.Sh EXAMPLES
.Bd -literal
# PASS ONLY logs with host from 8pm to 9am  (20:00:00 to 09:00:00)
# and also matching 'root' on logins
auth.info %regex -v -t '^1' %regex -m 'root' %classic /var/log/webserver
.Ed
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr syslog.conf 5 ,
.Xr om_classic 8 ,
.Xr om_mysql 8 ,
.Xr om_peo 8 ,
.Xr om_pgsql 8 ,
.Xr om_tcp 8 ,
.Xr om_udp 8 ,
.Xr peochk 8 ,
.Xr syslogd 8
.Sh BUGS
.Bl -bullet
.It
Still can't escape quotes within expression.
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/om_regex.c
/*	$CoreSDI: om_regex.c,v 1.13.2.10.4.17 2001/11/30 23:13:45 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * om_regex -- Filter messages using regular exressions
 *
 * Author: Alejo Sanchez for Core-SDI SA
 *         Idea of Emiliano Kargieman
 *
 */

#include "config.h"

#if TIME_WITH_SYS_TIME

# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <sys/param.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>
#include <regex.h>

#include "../modules.h"
#include "../syslogd.h"

/* current time from syslogd */

struct om_regex_ctx {
	short		flags;
	int		size;
	int		filters;
#define	OM_FILTER_MESSAGE	0x01
#define	OM_FILTER_HOST		0x02
#define	OM_FILTER_DATE		0x04
#define	OM_FILTER_TIME		0x08
#define	OM_FILTER_INVERSE	0x10
	regex_t		msg_exp;
	regex_t		host_exp;
	regex_t		date_exp;
	regex_t		time_exp;
};


/*
 *  INIT -- Initialize om_regex
 *
 */
int
om_regex_init(int argc, char **argv, struct filed *f, char *prog, void **ctx,
    char **status)
{
	struct	om_regex_ctx *c;
	regex_t	*creg;
	int	ch, statbuf_len;
	char	statbuf[1048], *p;

	creg = NULL;
	p = NULL;

	/* for debugging purposes */
	dprintf(MSYSLOG_INFORMATIVE, "om_regex init\n");

	if (argc < 2 || argv == NULL || argv[1] == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_regex: error on "
		    "initialization\n");
		return (-1);
	}

	if ((*ctx = (void *) calloc(1, sizeof(struct om_regex_ctx))) == NULL)
		return (-1);

	c = (struct om_regex_ctx *) *ctx;
	c->size = sizeof(struct om_regex_ctx);

	statbuf_len = snprintf(statbuf, sizeof(statbuf),
	    "om_regex: filtering");

	/*
	 * Parse options with getopt(3)
	 *
	 * we give an example for a -s argument
	 * -v flag means INVERSE matching
	 * -m flag match message
	 * -h flag match host
	 * -d flag match date
	 * -t flag match time
	 *
	 */
	p = NULL;
	optind = 1;
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	while ((ch = getopt(argc, argv, "vm:h:d:t:")) != -1) {
		switch (ch) {
		case 'v':
			c->filters |= OM_FILTER_INVERSE;
			statbuf_len += snprintf(statbuf + statbuf_len,
			    sizeof(statbuf) - statbuf_len, ", inverse");
			continue;

		case 'm':
			c->filters |= OM_FILTER_MESSAGE;
			creg = &c->msg_exp;
			p = ", message";
			break;

		case 'h':
			c->filters |= OM_FILTER_HOST;
			creg = &c->host_exp;
			p = ", host";
			break;

		case 'd':
			c->filters |= OM_FILTER_DATE;
			creg = &c->date_exp;
			p = ", date";
			break;

		case 't':
			c->filters |= OM_FILTER_TIME;
			creg = &c->time_exp;
			p = ", time";
			break;

		default:
			dprintf(MSYSLOG_SERIOUS, "om_regex: unknown parameter"
			    " [%c]\n", ch);
			free(*ctx);
			return (-1);
		}

		if (regcomp(creg, optarg, REG_EXTENDED | REG_NOSUB) != 0) {
			dprintf(MSYSLOG_SERIOUS, "om_regex: error compiling "
			    "regular expression [%s] for message\n", optarg);
			free(*ctx);
			return (-1);
		}

		if (p)
			statbuf_len += snprintf(statbuf + statbuf_len,
			    sizeof(statbuf) - statbuf_len, " %s [%s]", p,
			    optarg);
	}

	*status = strdup(statbuf);

	return (1);
}

/* return:
	 -1  error
	  1  match  -> successfull
	  0  nomatch -> stop logging it
 */

int
om_regex_write(struct filed *f, int flags, struct m_msg *m, void *ctx)
{
	struct om_regex_ctx *c;
	regex_t *creg;
	char *str, time_buf[16];
	int i, iflag;

	creg = NULL;
	str = NULL;

	c = (struct om_regex_ctx *) ctx;

	if (m == NULL || m->msg == NULL || !strcmp(m->msg, "")) {
		logerror("om_regex_write: no message!");
		return (-1);
	}

	/* Split date and time if filters are present. */
	if ((c->filters & OM_FILTER_DATE) || (c->filters & OM_FILTER_TIME)) {
		strftime(time_buf, sizeof(time_buf), "%b %e %H:%M:%S",
		    &f->f_tm);   

		time_buf[6] = 0;
		time_buf[15] = 0;
	}

	iflag = ((c->filters & OM_FILTER_INVERSE) != 0);

	for (i = 1; i < OM_FILTER_INVERSE; i <<= 1) {
		if ((c->filters & i) == 0)
			continue;

		switch (i) {
		case OM_FILTER_MESSAGE:
			creg = &c->msg_exp;
			str = m->msg;
			break;

		case OM_FILTER_HOST:
			creg = &c->host_exp;
			str = f->f_prevhost;
			break;

		case OM_FILTER_DATE:
			creg = &c->date_exp;
			str = time_buf;
			break;

		case OM_FILTER_TIME:
			creg = &c->time_exp;
			str = time_buf + 7;
			break;
		}

		if ((regexec(creg, str, 0, NULL, 0) != 0) ^ iflag)
				return (0);
	}

	return (1);

}

